File: TEMP.ViewModels/ChemicalTableViewModel.fs
namespace TEMP.ViewModels

open System.ComponentModel
open System.Collections.ObjectModel
open SafetyProgram.Core.Models
open SafetyProgram.UI.Views.ModelViews.ContentViews.DocumentObjects.ChemicalTables.Default

type ChemicalTableViewModel(chemicalTable : IChemicalTable) as this = 
    let mutable header = chemicalTable.Header
    let chemicals = chemicalTable.Content
    let propertyChangedEvent = new Event<_,_>()

    let setHeader newHeader = 
        header<-newHeader
        propertyChangedEvent.Trigger(
            this,
            new PropertyChangedEventArgs("Header"))

    do
        chemicalTable.HeaderChanged.Add(fun e ->
            e.NewProperty |> setHeader)

    interface IChemicalTableViewModel with
        member this.Header
            with get () = chemicalTable.Header
            and set value = setHeader value
        member this.Chemicals = chemicals
        [<CLIEvent>]
        member this.PropertyChanged = propertyChangedEvent.Publish

File: SafetyProgram.UI/ModelViews/ContentViews/DocumentObjects/ChemicalTables/Default/ChemicalTableViewModel.cs
using System.Collections.ObjectModel;
using System.ComponentModel;
using SafetyProgram.Base;
using SafetyProgram.Core.Models;

namespace SafetyProgram.UI.ModelViews.ContentViews.DocumentObjects.ChemicalTables.Default
{
    /// <summary>
    /// Defines a standard implementation of a IChemicalTableViewModel.
    /// </summary>
    internal sealed class ChemicalTableViewModel : IChemicalTableViewModel
    {
        /// <summary>
        /// Construct an instance of a ChemicalTableViewModel. A viewmodel for a chemical table view.
        /// </summary>
        /// <param name="chemicalTable">The underlying chemical table model to expose to the view.</param>
        /// <param name="chemicalTableCommands">Commands that act on the model.</param>
        public ChemicalTableViewModel(IChemicalTable chemicalTable)
        {
            this.chemicalTable = chemicalTable;

            this.Chemicals = chemicalTable.Content;

            chemicalTable.HeaderChanged +=
                (s, e) => PropertyChanged.Raise(this, "Header");
        }

        private IChemicalTable chemicalTable;

        /// <summary>
        /// Get or set the header of the chemical table.
        /// </summary>
        public string Header
        {
            get { return chemicalTable.Header; }
            set
            {
                chemicalTable.Header = value;
                PropertyChanged.Raise(this, "Header");
            }
        }

        /// <summary>
        /// Get the chemicals in the chemical table.
        /// </summary>
        public ObservableCollection<
            ICoshhChemical> Chemicals { get; private set; }

        /// <summary>
        /// Occurs when a property in this class changes.
        /// </summary>
        public event PropertyChangedEventHandler PropertyChanged;
    }
}
